library/ssl_tls.c

```diff
@@ -3203,7 +3203,7 @@ static void ssl_calc_finished_tls_sha384(
 {
     int len = 12;
     const char *sender;
-    unsigned char padbuf[48];
+    unsigned char padbuf[64];
```

Comment:
16 bytes isn't much, but it would be better to save them. This is one of those cases where we know better than the compiler. Could we pacify gcc with a cast (with a comment for the sake of humans)?

Reply:
It's hard do pacify GCC on this one.
1. I'm not allowed to cast to array types:
```c
mbedtls_sha512_finish_ret( &sha512, (unsigned char[64])padbuf );
```
Results in
`error: cast specifies array type`

2. Casting to a pointer triggers another warning (treated as error):
```c
mbedtls_sha512_finish_ret( &sha512, (unsigned char*)padbuf );
```
`‘mbedtls_sha512_finish_ret’ accessing 64 bytes in a region of size 48`.

---

I understand we have a few options:
1. Leave it as it is, and waste 16 bytes;
2. Change `mbedtls_sha512_finish_ret` to accept a pointer (loosing the benefits of "conformant arrays"). By the way, this function is marked for internal use only.
3. Disable a warning (I don't like this one);
4. Find another way to cast `padbuf`.

Any ideas?

Reply:
> this function is marked for internal use only.

Oh, that's an editorial mistake. This sentence was edited into the description of the wrong function in 27ff120a6121528de9f9a726dfd80a209ee05a1a. I'll make a PR to fix that.

Reply:
> Change `mbedtls_sha512_finish_ret` to accept a pointer

I'm not sure if that's acceptable. According to the C standard, a pointer and an array are equivalent in the declaration of a function parameter. But according to popular compilers, they aren't (which is why we're having this discussion).

Reply:
It's ugly, but how about casting the function?

Reply:
> It's ugly, but how about casting the function?

Maybe. Please, take a look --> d596ca8

Reply:
So casting to `int (*)(mbedtls_sha512_context*, unsigned char[48])` and not silences the warning? Wouldn't casting to `int (*)(mbedtls_sha512_context*, unsigned char*)` be more robust in case GCC's analysis starts going slightly deeper (which could happen due to an optimization flag)?

Reply:
It would be more generic, at least. I've just made this change.
We are not changing `mbedtls_sha512_finish_ret()` because it's part of the public API, right? Because I feel like the use of array parameters is not worth the trouble.
Is there an overall guideline for new public functions in this project (regarding array parameters)?

Reply:
Thanks, that's not ideal but I guess it's the best possible.

It hadn't been an issue before, but I think we'll make a new guideline to not use array parameters, and we'll change the existing ones to pointers in Mbed TLS 3.0.